### PR TITLE
GPR Update

### PR DESCRIPTION
**Pull request by hyunyong**
#### PR description:

GPR code update.
```GlobalTrackerMuonAlignment.cc```

Added some scripts for the GPR alignment.
```makeSub.py``` creates the HTcondor submit files.
```GlobalTrackerMuonAlignment_cfg.py``` calculates residuals.
```GlobalTrackerMuonAlignment_writeDB_cfg.py``` minimize and write DB.
#### PR validation:

This code is used for the 2022 dataset.

### CODE REVIEW

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 1]
```diff
@@ -108,108 +114,101 @@ using namespace reco;
...
+  static void fillDescriptions(edm::ConfigurationDescriptions &);
 
 private:
   void beginJob() override;
-  void analyze(const edm::Event&, const edm::EventSetup&) override;
+  void analyze(const edm::Event &, const edm::EventSetup &) override;
   void endJob() override;
 
   // ----------member data ---------------------------
-  const edm::ESGetToken<GlobalTrackingGeometry, GlobalTrackingGeometryRecord> m_TkGeometryToken;
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:159]: Why this change. This is undoing https://github.com/cms-sw/cmssw/commit/233abdec44ee94ce9036e6b731235593aa82e0bc

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 2]
```diff
@@ -19,84 +20,89 @@
...
 
-#include <iostream>
-#include <cmath>
 #include <cassert>
+#include <cmath>
 #include <fstream>
 #include <iomanip>
+#include <iostream>
+
+#include "FWCore/Framework/interface/EventSetup.h"  //MK Library
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:105]: what's this for? This is now duplicated.

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 3]
```diff
@@ -19,84 +20,89 @@
...
+#include "TrackingTools/PatternTools/interface/Trajectory.h"
 #include "TrackingTools/Records/interface/TrackingComponentsRecord.h"
 #include "TrackingTools/TrackRefitter/interface/TrackTransformer.h"
 #include "TrackingTools/TrajectoryState/interface/TrajectoryStateTransform.h"
-#include "TrackingTools/GeomPropagators/interface/Propagator.h"
-#include "TrackingTools/GeomPropagators/interface/SmartPropagator.h"
-#include "TrackingTools/GeomPropagators/interface/PropagationDirectionChooser.h"
 #include "TrackingTools/TransientTrack/interface/TransientTrack.h"
-#include "TrackPropagation/SteppingHelixPropagator/interface/SteppingHelixPropagator.h"
-#include "TrackPropagation/RungeKutta/interface/defaultRKPropagator.h"
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:69]: can you please alpha-order the includes and check for duplicates?

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 4]
```diff
@@ -108,108 +114,101 @@ using namespace reco;
...
 class GlobalTrackerMuonAlignment : public edm::one::EDAnalyzer<> {
 public:
-  explicit GlobalTrackerMuonAlignment(const edm::ParameterSet&);
-  ~GlobalTrackerMuonAlignment() override = default;
-  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
-
-  void analyzeTrackTrack(const edm::Event&, const edm::EventSetup&);
-  void analyzeTrackTrajectory(const edm::Event&, const edm::EventSetup&);
+  explicit GlobalTrackerMuonAlignment(const edm::ParameterSet &);
+  ~GlobalTrackerMuonAlignment() override;
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:118]: what's the point of this change?

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 5]
```diff
@@ -237,147 +246,123 @@ class GlobalTrackerMuonAlignment : public edm::one::EDAnalyzer<> {
...
-  //GlobalTrackerMuonAlignment::analyzeTrackTrack(iEvent, iSetup);
+void GlobalTrackerMuonAlignment::analyze(const edm::Event &iEvent, const edm::EventSetup &iSetup) {
   GlobalTrackerMuonAlignment::analyzeTrackTrajectory(iEvent, iSetup);
 }
 
-// ------------ method called once for job just before starting event loop  ------------
+// ------------ method called once for job just before starting event loop
+// ------------
 void GlobalTrackerMuonAlignment::beginJob() {
+  std::cout << "beginJob start" << std::endl;
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:362]: no `cout`s.

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 6]
```diff
@@ -237,147 +246,123 @@ class GlobalTrackerMuonAlignment : public edm::one::EDAnalyzer<> {
...
+  desc.add<InputTag>("smuons", edm::InputTag("ALCARECOMuAlCalIsolatedMu:SelectedMuons"));
+  desc.add<InputTag>("ref_track", edm::InputTag("MuonAlignmentFromReferenceGlobalMuonRefit:Refitted"));
+  desc.add<InputTag>("ref_muon", edm::InputTag("MuonAlignmentFromReferenceGlobalMuonRefit2:Refitted"));
+  descriptions.add("GlobalTrackerMuonAlignment", desc);
 }
-//
-// member functions
-//
+
+GlobalTrackerMuonAlignment::~GlobalTrackerMuonAlignment() {}
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:352]: if it's empty it can stay `default`.

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 7]
```diff
@@ -405,9 +400,20 @@ void GlobalTrackerMuonAlignment::beginJob() {
   dirsave->cd();
 }
 
-// ------------ method called once each job just after ending the event loop  ------------
+// ------------ method called once each job just after ending the event loop
+// ------------
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:404]: ?

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 8]
```diff
@@ -422,8 +428,7 @@ void GlobalTrackerMuonAlignment::endJob() {
     }
   bool ierr = !InfI.Invert();
   if (ierr) {
-    if (alarm)
-      edm::LogVerbatim("GlobalTrackerMuonAlignment") << " Error inverse  Inf matrix !!!!!!!!!!!";
+    std::cout << " Error inverse  Inf matrix !!!!!!!!!!!" << std::endl;
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:431]: no `cout`s.

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc [hunk 9]
```diff
@@ -436,8 +441,7 @@ void GlobalTrackerMuonAlignment::endJob() {
   int iEr3;
   CLHEP::HepMatrix Errd3 = Hess.inverse(iEr3);
   if (iEr3 != 0) {
-    if (alarm)
-      edm::LogVerbatim("GlobalTrackerMuonAlignment") << " endJob Error inverse Hess matrix !!!!!!!!!!!";
+    std::cout << " endJob Error inverse Hess matrix !!!!!!!!!!!" << std::endl;
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:444]: no `cout`s.

**Review by mmusich — CHANGES_REQUESTED** [review]
@hyunyong it looks like you started from an older version of this file, undoing all the maintenance work of the last couple of years. Please re-apply your changes to the latest version of this file.

**Comment by tvami** [comment]
hey @hyunyong do you have an update on this?

**Comment by tvami** [comment]
@hyunyong please do code checks and squash your commits into one

File: Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc (hunk 4 above)
Comment (tvami) [reply Alignment/CommonAlignmentProducer/plugins/GlobalTrackerMuonAlignment.cc:110]: I see this was still changed to `override`

**Comment by hyunyong** [comment]
> @hyunyong please do code checks and squash your commits into one

The code is based on a pretty old one. I will compare the current CMSSW version.

**Comment by tvami** [comment]
OK, please rebase to 13_1_X

**Comment by tvami** [comment]
@hyunyong just a kind ping :)

**Comment by tvami** [comment]
hi @hyunyong just a reminder about this PR

**Comment by tvami** [comment]
hi @hyunyong , any updates on this?

**Comment by hyunyong** [comment]
Hi @tvami,

We are developing a new global alignment method that might not use the GPR tag anymore.
The current GPR method assumes a slight shift of muon individual chambers and affects the final alignment results. I will report details of it soon and try to use it for the 2022 end-of-year alignment and 2023 rereco.

**Comment by tvami** [comment]
Sounds great! Will you close this PR then?

**Comment by hyunyong** [comment]
> Sounds great! Will you close this PR then?

Ok, I close it.
